avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
         :param expire: time in seconds for the asset to expire
         :param metadata: metadata which will be saved inside metadata file
         """
-        self.name = name
+        parsed_url = urlparse.urlparse(name)
+        # Asset file name
+        self.basename = os.path.basename(parsed_url.path)
         self.asset_hash = asset_hash
-        if algorithm is None:
-            self.algorithm = DEFAULT_HASH_ALGORITHM
-        else:
+
+        if algorithm:
             self.algorithm = algorithm
+        else:
+            self.algorithm = DEFAULT_HASH_ALGORITHM
 
+        self.locations = []
+        # If `name` is an url, move it to `locations`
+        if parsed_url.scheme:
+            self.locations.append(os.path.dirname(parsed_url.geturl()))
+        # Let's be conservative and check `locations` type
         if isinstance(locations, str):
-            self.locations = [locations]
+            # remove asset name from location
+            if self.basename in locations:
+                parsed_locations = urlparse.urlparse(locations)
+                self.locations.append(
+                    os.path.dirname(parsed_locations.geturl()))
```

Comment:
Suppose `locations='http://example2.org/foo/bar'`  and `name=file.tgz`.

`os.path.dirname(parsed_locations.geturl())` results in `http://example2.org/foo/`. But don't you want the location `http://example2.org/foo/bar`?

Reply:
I ran this code:
```
>>> locations='http://example2.org/foo/bar'
>>> parsed_locations = urlparse.urlparse(locations)
>>> os.path.dirname(parsed_locations.geturl())
'http://example2.org/foo'
```

Reply:
This would be true if line 91 was not there. Lines 92 through 94 will be executed just if we have something like `self.basename='file.tgz'` and `locations='http://example2.org/foo/bar/file.tgz'`.

This block of code is just to make sure we keep the asset name and its location separate and in their own places.